### ISSUE TITLE
v241 has issues with /home encrypted

### ISSUE THREAD

**Issue by hphilm** [systemd/systemd#11723]
**systemd version the issue has been seen with**
> v241

**Used distribution**
 > Manjaro Linux KDE 18.0.3-pre2

**Expected behaviour you didn't see**
 > unencrypted `/home` partition

**Unexpected behaviour you saw**
 > manjaro-kde systemd-cryptsetup[269]: Failed to open key file.
manjaro-kde systemd-cryptsetup[269]: Failed to activate with the key file '/crypto_keyfile.bin' : Invalid argument
manjaro-kde systemd[1]: Failed to start Cryptography Setup for luks-[UUID]

**Steps to reproduce the problem**

Installed a KDE VM system using [latest development ISO](https://osdn.net/projects/manjaro/storage/kde/18.0.3-pre2/), manually partitioned with efi boot partition, root partition and separate encrypted home partition.

Reported on Manjaro forum:

https://forum.manjaro.org/t/75655/26
https://forum.manjaro.org/t/75711

Issue is most likely introduced via [23th Jan changes](https://github.com/systemd/systemd/compare/v241-rc1...v241-rc2) to cryptsetup.

**Comment by philmmanjaro** [comment]
Seems reverting 6f177c7dc092eb68762b4533d41b14244adb2a73 fixes the issue. So we have to revisit that commit once more ...

**Comment by yuwata** [comment]
Could you show /etc/crypttab and generated `systemd-cryptsetup@xxx.service`?

**Comment by poettering** [comment]
So, that file /crypto_keyfile.bin does it actually contain a valid key?

Previously, if we get EINVAL from an attempt to activate some disk while passing a key file, we'd try again with a password prompt to the user. Now we only do that if we get EPERM (which is the usual error the kernel returns if the key is not correct).

Hence, does /crypto_keyfile.bin actually work? is that file valid? or did you just specify it but enter the password manually anyway?

**Comment by philmmanjaro** [comment]
Since I don't use encryption myself I've to ask my community. As soon as I've some feedback I might provide it here.

**Comment by FrogMcRibbit** [comment]
I reproduced the issue with the following ISO: https://osdn.net/projects/manjaro/storage/kde/18.0.3-pre2/
I simply installed it in a virtual machine (Virtual Box 6.0.4, doesn't use EFI) with the following partitioning (made manually with Calamares):
- Unencrypted /
- Encrypted /home with LUKS
- MBR partition table (not GPT)

Fortunately, I still have access to CLI with the root user. I could retrieved the file you asked by mounting a USB drive in the vm and mount it. I'll put it in a zip file, along with some additional files that might help to understand what happened (a screenshot, some `ls` output and also the output given by `journalctl -xb`). All additional files have ADDED- in their name.

[issue11723-files-frog.zip](https://github.com/systemd/systemd/files/2877574/issue11723-files-frog.zip)

> So, that file /crypto_keyfile.bin does it actually contain a valid key?

I do not have `/crypto_keyfile.bin`; normally, I manually enter my passphrase at each boot.

**Comment by TotallyNotElite** [comment]
> So, that file /crypto_keyfile.bin does it actually contain a valid key?

I don't have that file, but it doesn't seem to matter. The system doesn't boot with systemd with 6f177c7, but as soon as I update to a version without 6f177c7 the system is able to boot just fine after I enter the password. No reinstall required or anything like that.

Here's my /etc/crypttab file. The file crypto_keyfile.bin does not exist but is still mentioned in the file for some reason. The password used was "manjaro".
![image](https://user-images.githubusercontent.com/38938720/52997618-5835d980-3421-11e9-8cd6-db00d4c3c30d.png)

**Comment by philmmanjaro** [comment]
Seems there is a regression with **cryptsetup v2.1.0**. I'll push a new **systemd** without the revert and **cryptsetup v2.0.6** to our **unstable** branch. @TotallyNotElite and @FrogMcRibbit: may you test that combination when you use **Manjaro**? Also this affects somehow our install medias **v18.0.3** which are not able to encrypt at all thru **Calamares** installer ...

**Comment by philmmanjaro** [comment]
After testing with the commit https://github.com/systemd/systemd/commit/6f177c7dc092eb68762b4533d41b14244adb2a73 not reverted and **cryptsetup v2.0.6** I got the same error as reported already. So **cryptsetup** doesn't matter in this issue here ...

**Comment by TotallyNotElite** [comment]
> After testing with the commit [6f177c7](https://github.com/systemd/systemd/commit/6f177c7dc092eb68762b4533d41b14244adb2a73) not reverted and **cryptsetup v2.0.6** I got the same error as reported already. So **cryptsetup** doesn't matter in this issue here ...

I suspect the issue is caused when systemd/cryptsetup can't find the /crypto_keyfile.bin file (which is mentioned in /etc/crypttab) and doesn't ask the user for a password for some reason.

**Comment by kirbyfan64** [comment]
@TotallyNotElite Just to check, can you try changing the `/crypto_keyfile.bin` field to `-` and see if it works?

**Comment by kirbyfan64** [comment]
Some more information: based on the given log messages, [cryptsetup is the one failing](https://github.com/mbroz/cryptsetup/blob/a6f5ce8c7bfeff945f6a87ca1791ad5b9dae145b/lib/utils.c#L170) and also swallows the errno result. 

I would personally argue that systemd should always ask for a password if the keyfile reading fails (e.g. treat it as it does EAGAIN), since IMO it would be a good fallback for the file being missing or corrupted. 

As for how a nonexistent keyfile ended up in the crypttab, I suspect something to do with Manjaro, maybe its installer? That would explain why the only other current cases of this bug are from Manjaro users.

**Comment by FrogMcRibbit** [comment]
I confirm that removing the reference to /crypto_keyfile.bin file completely fix the issue. I just removed the field and left it empty, which is supposed to be equivalent according to [crypttab(5)](https://www.freedesktop.org/software/systemd/man/crypttab.html) manual.

Someone else on Manjaro forum (thanks sueridgepipe) confirm this too. See his report here: https://forum.manjaro.org/t/unstable-update-2019-02-22-systemd-cryptsetup-go-python-might-break-encrypted-systems-with-luks/76767/11

I think this is a regression, and that it should definitely fallback to asking to type the password manually if the file referenced in `/etc/crypttab` doesn't exist. Not doing so mean that if anything wrong happens to the file and got moved or deleted, if the `/etc/crypttab` is not modified, the user will be forced to troubleshoot immediately. Also, all people that are using encrypted partitions configured that way (most likely by the installer of a distro) will face a serious regression when upgrading to systemd 241 or higher if nothing change.

It worked perfectly fine before, for example on systemd 239. It also works fine with 241 + revert https://github.com/systemd/systemd/commit/6f177c7dc092eb68762b4533d41b14244adb2a73, which is what Manjaro did [as a temporary measure](https://gitlab.manjaro.org/packages/core/systemd/blob/master/PKGBUILD).
